hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/OMDBInsightEndpoint.java

```diff
@@ -0,0 +1,412 @@
+ * delete files in FSO buckets.
+ * 5. Amount of data mapped to pending delete keys in legacy/OBS buckets and
+ * pending delete files in FSO buckets.
+ */
+@Path("/omdbinsight")
+@Produces(MediaType.APPLICATION_JSON)
+@AdminOnly
+public class OMDBInsightEndpoint {
+
+  @Inject
+  private ContainerEndpoint containerEndpoint;
+  @Inject
+  private ReconContainerMetadataManager reconContainerMetadataManager;
+  private final ReconOMMetadataManager omMetadataManager;
+  private final ReconContainerManager containerManager;
+
+  @Inject
+  public OMDBInsightEndpoint(OzoneStorageContainerManager reconSCM,
+                             ReconOMMetadataManager omMetadataManager) {
+    this.containerManager =
+        (ReconContainerManager) reconSCM.getContainerManager();
+    this.omMetadataManager = omMetadataManager;
+  }
+
+  /**
+   * This method retrieves set of keys/files which are open.
+   *
+   * @return the http json response wrapped in below format:
+   * {
+   *     replicatedTotal: 13824,
+   *     unreplicatedTotal: 4608,
+   *     entities: [
+   *     {
+   *         path: “/vol1/bucket1/key1”,
+   *         keyState: “Open”,
+   *         inStateSince: 1667564193026,
+   *         size: 1024,
+   *         replicatedSize: 3072,
+   *         unreplicatedSize: 1024,
+   *         replicationType: RATIS,
+   *         replicationFactor: THREE
+   *     }.
+   *    {
+   *         path: “/vol1/bucket1/key2”,
+   *         keyState: “Open”,
+   *         inStateSince: 1667564193026,
+   *         size: 512,
+   *         replicatedSize: 1536,
+   *         unreplicatedSize: 512,
+   *         replicationType: RATIS,
+   *         replicationFactor: THREE
+   *     }.
+   *     {
+   *         path: “/vol1/fso-bucket/dir1/file1”,
+   *         keyState: “Open”,
+   *         inStateSince: 1667564193026,
+   *         size: 1024,
+   *         replicatedSize: 3072,
+   *         unreplicatedSize: 1024,
+   *         replicationType: RATIS,
+   *         replicationFactor: THREE
+   *     }.
+   *     {
+   *         path: “/vol1/fso-bucket/dir1/dir2/file2”,
+   *         keyState: “Open”,
+   *         inStateSince: 1667564193026,
+   *         size: 2048,
+   *         replicatedSize: 6144,
+   *         unreplicatedSize: 2048,
+   *         replicationType: RATIS,
+   *         replicationFactor: THREE
+   *     }
+   *   ]
+   * }
+   */
+  @GET
+  @Path("openkeyinfo")
+  public Response getOpenKeyInfo(
+      @DefaultValue(DEFAULT_FETCH_COUNT) @QueryParam(RECON_QUERY_LIMIT)
+      int limit,
+      @DefaultValue(StringUtils.EMPTY) @QueryParam(RECON_QUERY_PREVKEY)
+      String prevKeyPrefix) {
+    KeyInsightInfoResp openKeyInsightInfo = new KeyInsightInfoResp();
+    List<KeyEntityInfo> nonFSOKeyInfoList =
+        openKeyInsightInfo.getNonFSOKeyInfoList();
+    boolean isLegacyBucketLayout = true;
+    boolean recordsFetchedLimitReached = false;
+    List<KeyEntityInfo> fsoKeyInfoList = openKeyInsightInfo.getFsoKeyInfoList();
```

Comment:
The logic for both Opentable and deleteTable seems similar with Table fetch, using iterator  , creating KeyEntity and adding to Response and updating  overall un replication size  and replication size.
Is there a way to optimise and use common code for Table iterator and updates for Open as well deleted case.
Also for OMkeyInfo is transformed to KeyEntity object for Response in case on OpenKey as well as DeletedDir , but for deletedKey repeatedOmKeyInfo is directly returned which has OMkeyInfo List .
 So not sure about this logic but either in other case we could  OMkeyInfo or also transform  repeatedOmKeyInfo to Keyentity List

Reply:
@krishnaasawa1  thanks for your review, so on you below point:
1. Is there a way to optimise and use common code for Table iterator and updates for Open as well deleted case. 
      -  Both Open Key and Deleted key responses are different , though they are prepared after iteration from their respective tables. In openKey API response, logic is different as we have to consider both OBS/LEGACY and FSO buckets because the tsandard interface provided by omMetadataManager accepts bucketlayout and provides the reference of "openKeyTable" or "openFileTable" respectively and their iterators return "Table<String, OmKeyInfo>" , however in case of deletedKeys API, omMetadataManager doesn't have any standard interface for two different tables ("deletedTable" and "deletedDirectoryTable" ), so we don't have to provide bucket layout as an argument and we have to fetch data from "deletedTable" for keys and need to fetch data from "deletedDirectoryTable" for directories which is done separately in two different methods due to their tables provide different iterators like "Table<String, RepeatedOmKeyInfo>" and "Table.KeyValue<String, OmKeyInfo>>", so you can see here, iterator logic cannot be same. because one provides RepeatedKeyInfo and another just single OmKeyInfo. And that is the reason we have different response for different APIs as per design doc we discussed.

Reply:
Thanks for detailed response. It's okay to keep same if we can't optimize.